Stop abusing Twine in DiagnosticInfo

bogner opened the pull request:
This switches the various DiagnosticInfo subclasses that store an un-owned string to use StringRef for this instead of Twine.

Storing a `Twine &` like most of these were doing makes it very easy to introduce memory corruption bugs:

```c++
void f(std::string &Msg) {
  auto Diag = DiagnosticInfoGeneric(Msg);
  Ctx.diagnose(Diag); // Bug! We're referring to the temporary Twine!
  Ctx.diagnose(DiagnosticInfoGeneric(Msg); // this works, I guess...
};
```

This was sort of mitigated in DiagnosticInfoUnsupported, which actually copies the Twine, but copying Twines is always a questionable practice.

Instead, we store a StringRef, which makes it much more explicit that ownership of the storage is the caller's problem.

bogner committed:
Stop abusing Twine in DiagnosticInfo

This switches the various DiagnosticInfo subclasses that store an un-owned
string to use StringRef for this instead of Twine.

Storing a `Twine &` like most of these were doing makes it very easy to
introduce memory corruption bugs:

```c++
void f(std::string &Msg) {
  auto Diag = DiagnosticInfoGeneric(Msg);
  Ctx.diagnose(Diag); // Bug! We're referring to the temporary Twine!
  Ctx.diagnose(DiagnosticInfoGeneric(Msg); // this works, I guess...
};
```

This was sort of mitigated in DiagnosticInfoUnsupported, which actually copies
the Twine, but copying Twines is always a questionable practice.

Instead, we store a StringRef, which makes it much more explicit that ownership
of the storage is the caller's problem.

arsenm reviewed (changes requested):
Can we just get a warning on the bad usage instead?

arsenm reviewed (commented):
If we do this the DiagnosticPrinter twine overload of << is probably dead?

bogner commented:
> Can we just get a warning on the bad usage instead?

I'd be perfectly happy for this to just warn on misuse, but I don't think it's possible. Essentially we'd need to be able to warn on creating any of these `DiagnosticInfo*` classes on the stack, as the problematic pattern is as follows:
```c++
DiagnosticInfo Foo(...);
Ctx.diagnose(Foo);
```


The only way to correctly use these diagnostic classes is within the expression that calls `LLVMContext::diagnose`. This is the same as the problem with `Twine` itself, where we also need to ensure we're never doing this. IIRC we've discussed creating a warning for that too, but nobody came up with anything satisfactory. The only real difference is really that it's well known how `Twine` works and where it's dangerous, so people tend to catch misuses in code review.

If we want to make these interfaces "safe", we either need to move them away from `Twine`, or make them actually copy and store the string. If we *don't* want to make these safe because we care about the miniscule amount of extra stack space this approach uses in error paths, then I guess this isn't necessary and we can just continue to live with watching people debug the subtle build configuration and platform specific bugs that occur with these types.

bogner committed:
fixup: Fix target specific diags that were missed

bogner commented:
> If we do this the DiagnosticPrinter twine overload of << is probably dead?

It should be, yes. Updated the patch to catch the target specific diags I missed that removing this caught.

bogner committed:
fixup: clean up uses where the string *was* copied

These uses were fine, but the comment that claimed the string isn't
copied are just incorrect.

bogner committed:
fixup: adjacent clang-format to appease the bot

bogner committed:
fixup: typos

arsenm commented:
> > Can we just get a warning on the bad usage instead?
> 
> I'd be perfectly happy for this to just warn on misuse, but I don't think it's possible. Essentially we'd need to be able to warn on creating any of these `DiagnosticInfo*` classes on the stack, as the problematic pattern is as follows:

Seems to work if we add `[[clang::lifetimebound]]` to all of these constructor parameters 

```
diff --git a/llvm/include/llvm/IR/DiagnosticInfo.h b/llvm/include/llvm/IR/DiagnosticInfo.h
index 8743f8058c38..badc1cdb6cac 100644
--- a/llvm/include/llvm/IR/DiagnosticInfo.h
+++ b/llvm/include/llvm/IR/DiagnosticInfo.h
@@ -146,15 +146,15 @@ public:
   /// \p MsgStr is the message to be reported to the frontend.
   /// This class does not copy \p MsgStr, therefore the reference must be valid
   /// for the whole life time of the Diagnostic.
-  DiagnosticInfoGeneric(const Twine &MsgStr,
+  DiagnosticInfoGeneric(const Twine &MsgStr [[clang::lifetimebound]] ,
                         DiagnosticSeverity Severity = DS_Error)
       : DiagnosticInfo(DK_Generic, Severity), MsgStr(MsgStr) {}
 
-  DiagnosticInfoGeneric(const Instruction *I, const Twine &ErrMsg,
+  DiagnosticInfoGeneric(const Instruction *I, const Twine &ErrMsg [[clang::lifetimebound]] ,
                         DiagnosticSeverity Severity = DS_Error)
       : DiagnosticInfo(DK_Generic, Severity), MsgStr(ErrMsg), Inst(I) {}
 
-  const Twine &getMsgStr() const { return MsgStr; }
+  const Twine &getMsgStr() const [[clang::lifetimebound]] { return MsgStr; }
   const Instruction *getInstruction() const { return Inst; }
 
   /// \see DiagnosticInfo::print.
@@ -193,7 +193,7 @@ public:
                           DiagnosticSeverity Severity = DS_Error);
 
   uint64_t getLocCookie() const { return LocCookie; }
-  const Twine &getMsgStr() const { return MsgStr; }
+  const Twine &getMsgStr() const [[clang::lifetimebound]] { return MsgStr; }
   const Instruction *getInstruction() const { return Instr; }
 
   /// \see DiagnosticInfo::print.
@@ -364,7 +364,7 @@ public:
 
   /// Return the absolute path tot the file.
   std::string getAbsolutePath() const;
-  
+
   const Function &getFunction() const { return Fn; }
   DiagnosticLocation getLocation() const { return Loc; }
 
@@ -391,7 +391,7 @@ public:
       : DiagnosticInfoWithLocationBase(DK_GenericWithLoc, Severity, Fn, Loc),
         MsgStr(MsgStr) {}
 
-  const Twine &getMsgStr() const { return MsgStr; }
+  const Twine &getMsgStr() const [[clang::lifetimebound]] { return MsgStr; }
 
   /// \see DiagnosticInfo::print.
   void print(DiagnosticPrinter &DP) const override;
@@ -417,7 +417,7 @@ public:
   DiagnosticInfoRegAllocFailure(const Twine &MsgStr, const Function &Fn,
                                 DiagnosticSeverity Severity = DS_Error);
 
-  const Twine &getMsgStr() const { return MsgStr; }
+  const Twine &getMsgStr() const [[clang::lifetimebound]] { return MsgStr; }
 
   /// \see DiagnosticInfo::print.
   void print(DiagnosticPrinter &DP) const override;

```
```
 warning: temporary whose address is used as value of local variable 'Diag' will be destroyed at the end of the full-expression [-Wdangling]
 2863 |     auto Diag = DiagnosticInfoGeneric("bug");
```

bogner commented:
> > > Can we just get a warning on the bad usage instead?
> > 
> > 
> > I'd be perfectly happy for this to just warn on misuse, but I don't think it's possible. Essentially we'd need to be able to warn on creating any of these `DiagnosticInfo*` classes on the stack, as the problematic pattern is as follows:
> 
> Seems to work if we add `[[clang::lifetimebound]]` to all of these constructor parameters

This looks promising, I’ll look into doing that instead.

bogner commented:
Superceded by #137397, which opts to warn on misuse instead.
